config/routes/spree.rb

```diff
@@ -94,6 +94,16 @@
   resources :states, :only => :index
   resources :countries, :only => :index
 
+  # For some unknown reason this route /orders/populate and it's redirect
+  #   make product import routes work in a form in
+  #   app/views/admin/product_import/import.html.haml
+  #.  This route is otherwise dead
+  populate_redirect = redirect do |params, request|
+    request.flash[:error] = Spree.t(:populate_get_error)
+    request.referer || '/cart'
+  end
+  get '/orders/populate', :via => :get, :to => populate_redirect
```

Comment:
Which spec failed? I just checked out your pull request without this commit and `spec/features/admin/product_import_spec.rb` passed. Or do you actually remove the frontend gem for testing?

Reply:
I run the build against a version of spree where I simply deleted the folder frontend. The branch is here:
https://github.com/luisramos0/spree/tree/2-0-4-without-frontend

You can try it by changing the gemfile:
gem 'spree', github: 'luisramos0/spree', branch: '2-0-4-without-frontend'

Reply:
I didnt see the problem again! so I removed the commit with the populate route :-D
I did use different versions of the spree branch in the process: removing just the views first, then the controllers, maybe that caused the problem. With this last branch (no frontend folder) the build is green without the populate route.
thanks for asking :-D

Reply:
because you have suggested removing the gem I went ahead and tried replacing the spree dependency with individual dependencies to the different spree gems. it worked! the build is green, all looks good! #exciting

Reply:
in this PR I am separating spree the different spree gems and all works fine.

in the final PR #3724 I remove the spree_frontend gem and get the same product import spec errors that are fixed by adding the order/populate routes, you can have a look here @mkllnk 
https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/branches/pull-request-3724/builds/2

Reply:
Interesting. Yesterday, when I started a local server and requested the home page for the first time, the little profiler in the top left corner displayed a product import request that took really long. I couldn't reproduce it and I have no idea how that is possible but we will find out if it's a real problem.

Reply:
I can reproduce it now. Thank you. I still don't understand it though. If I pry into the view and run the failing spec I can issue these commands:
```
> url_for(:controller=>"admin/product_import", :action=>"import")
ActionController::RoutingError: No route matches {:controller=>"admin/product_import", :action=>"import"}
from /home/maikel/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/actionpack-3.2.22.5/lib/action_dispatch/routing/route_set.rb:540:in `raise_routing_error'
Caused by Journey::Router::RoutingError: Journey::Router::RoutingError
from /home/maikel/.rbenv/versions/2.1.5/lib/ruby/gems/2.1.0/gems/journey-1.0.4/lib/journey/formatter.rb:46:in `generate'

> main_app.url_for(:controller=>"admin/product_import", :action=>"import")
=> "http://127.0.0.1:44929/admin/product_import"

> Rails.application.routes.url_for(:controller=>"admin/product_import", :action=>"import")
=> "http://test.host/admin/product_import"
```

The route is clearly there and can be found. But the current scope seems to have a `url_for` method that is looking in the wrong place.

```
> method(:url_for)
=> #<Method: #<Class:0x00562666a25de0>(ActionView::Helpers::UrlHelper)#url_for>
```

That method is calling super. But where is that code? I pry in there and then:

```
> self.class.ancestors.select { |c| c.method_defined?(:url_for) }
=> [#<Class:0x00562666a25de0>, ActionView::Base, ActionView::Helpers::UrlHelper, ActionDispatch::Routing::UrlFor]
```

The first class is `self.class`. So the `super` call goes to `ActionView::Base but I can't find a definition in there. Just this interesting piece of code:

```rb
      # This method receives routes and helpers from the controller
      # and return a subclass ready to be used as view context.
      def prepare(routes, helpers) #:nodoc:
        Class.new(self) do
          if routes
            include routes.url_helpers
            include routes.mounted_helpers
          end

          if helpers
            include helpers
            self.helpers = helpers
          end
        end
      end
```

Calling `ActionView::Base.new.url_for` leads to `ActionDispatch::Routing::UrlFor`:
```rb
          _routes.url_for((options || {}).symbolize_keys.reverse_merge!(url_options))
```

That `_routes` is provided by the controller:

```
> Admin::ProductImportController.ancestors.select { |c| c.method_defined?(:_routes) }
=> [Admin::ProductImportController,
 Spree::Admin::BaseController,
 Spree::BaseController,
 #<Module:0x0056265fd5e5b8>,
 ApplicationController,
 #<Module:0x005626627f85f8>,
 ActionController::Base,
 AbstractController::UrlFor]
```

And there we go:
```
> Spree::Admin::BaseController._routes.url_for(:controller=>"admin/product_import", :action=>"import")
ActionController::RoutingError: No route matches {:controller=>"admin/product_import", :action=>"import"}
```

So that Spree controller seems to contain only Spree routes while the ApplicationController has our routes as well:
```
> ApplicationController._routes.routes.map do |route|   {alias: route.name, path: route.path.spec.to_s, controller: route.defaults[:controller], action: route.defaults[:action]} end.select { |a| a[:path].starts_with? "/admin/product_import" }
=> [{:alias=>"admin_product_import", :path=>"/admin/product_import(.:format)", :controller=>"admin/product_import", :action=>"index"},
 {:alias=>nil, :path=>"/admin/product_import(.:format)", :controller=>"admin/product_import", :action=>"import"},
 {:alias=>"admin_product_import_process_async", :path=>"/admin/product_import/validate_data(.:format)", :controller=>"admin/product_import", :action=>"validate_data"},
 {:alias=>"admin_product_import_save_async", :path=>"/admin/product_import/save_data(.:format)", :controller=>"admin/product_import", :action=>"save_data"},
 {:alias=>"admin_product_import_reset_async", :path=>"/admin/product_import/reset_absent(.:format)", :controller=>"admin/product_import", :action=>"reset_absent_products"}]

> Spree::Admin::BaseController._routes.routes.map do |route|   {alias: route.name, path: route.path.spec.to_s, controller: route.defaults[:controller], action: route.defaults[:action]} end.select { |a| a[:path].starts_with? "/admin/product_import" }
=> []
```

So it looks like removing the Spree frontend changes the loading of routes. I also noticed that putting in the orders/populate route adds it to the overview in `rake routes` three times, in different places. It seems that it also triggers OFN routes to appear in the Spree admin routes. There are more controllers inheriting from `Spree::Admin::BaseController` which could be affected by this:

```
app/controllers/admin/bulk_line_items_controller.rb
app/controllers/admin/cache_settings_controller.rb
app/controllers/admin/contents_controller.rb
app/controllers/admin/invoice_settings_controller.rb
app/controllers/admin/manager_invitations_controller.rb
app/controllers/admin/matomo_settings_controller.rb
app/controllers/admin/product_import_controller.rb
app/controllers/admin/stripe_accounts_controller.rb
app/controllers/admin/stripe_connect_settings_controller.rb
app/controllers/spree/admin/base_controller_decorator.rb
app/controllers/spree/admin/invoices_controller.rb
```

For example:
```
> ApplicationController._routes.url_for(:controller=>"admin/bulk_line_items", :action=>"index")
=> "http://test.host/admin/bulk_line_items"

> Spree::Admin::BaseController._routes.url_for(:controller=>"admin/bulk_line_items", :action=>"index")
ActionController::RoutingError: No route matches {:controller=>"admin/bulk_line_items"}
```

I feel a bit uneasy to solve this problem by defining a random route without understanding how it works. Do you think you could find out more?

Reply:
Dr Maikel the detective! Amazing. It would take me ages to get to this detail... and I am afraid it will take me ages fo figure this out. But yes, I think I need to have a look...

Reply:
ah, I found the problem! it's not random.

form_tag false... in product import view is generating this html:
![Screenshot 2019-05-28 at 12 29 28](https://user-images.githubusercontent.com/1640378/58474984-9d665780-8144-11e9-8e4c-964da7527bb1.png)

so, it's calling the orders_populate route.
I dont understand why the orders populate route is picked up here, but it doesnt matter, we can just fix it by setting the correct action in this form which is simply
form_tag main_app.admin_product_import_path
I have added 2 commits from #3724 in this PR and added this new one. I will close #3724 now.
ready for review!

Reply:
Wow! I'm so happy that you found the cause! :tada: :man_dancing: